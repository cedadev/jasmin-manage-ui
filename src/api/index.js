import { useEffect, useState } from 'react';

import {
    apiFetch,
    createContextForEndpoint,
    createContextForResource,
    useResource,
    useInstance,
    useFetchPoint
} from '../rest-resource';


// Create a context for the current user and export the hook
const CurrentUser = createContextForEndpoint("/api/me/");
const CurrentUserProvider = CurrentUser.Provider;
export const useCurrentUser = CurrentUser.hook;


// All other resources require authentication before they can be fetched
// This higher-order component wraps a provider so that it is only mounted once
// the user is successfully initialised
const withAuthentication = Provider => ({ children }) => {
    const currentUser = useCurrentUser();
    // Only wrap the children in the provider if the user is authenticated
    return currentUser.initialised ? <Provider>{children}</Provider> : children;
};


// Create a context each for the top-level, read-only resources that don't change
// often and we want to share across the component tree
const Categories = createContextForResource("/api/categories/");
const Consortia = createContextForResource("/api/consortia/");
const Resources = createContextForResource("/api/resources/");

// Wrap the providers with the higher-order component for authentication
const CategoriesProvider = withAuthentication(Categories.Provider);
const ConsortiaProvider = withAuthentication(Consortia.Provider);
const ResourcesProvider = withAuthentication(Resources.Provider);

// Export the hooks for each resource
export const useCategories = Categories.hook;
export const useConsortia = Consortia.hook;
export const useResources = Resources.hook;

// Export a single provider that provides all the top-level resources
export const Provider = ({ children }) => (
    <CurrentUserProvider>
        <CategoriesProvider>
            <ConsortiaProvider>
                <ResourcesProvider>
                    {children}
                </ResourcesProvider>
            </ConsortiaProvider>
        </CategoriesProvider>
    </CurrentUserProvider>
);


// Export a hook for loading a consortium by id
export const useConsortium = (id, options) => useInstance(`/api/consortia/${id}/`, options);

// Export hooks for accessing projects
// We don't treat the project list as a top-level shared resource because:
//   1. The list, or objects within it, changes relatively often
//   2. Projects can be loaded by id that are not in the list, e.g. when the user
//      is a consortium manager, which reduces the value in caching the list
export const useProjects = options => useResource("/api/projects/", options);
export const useProject = (id, options) => useInstance(`/api/projects/${id}/`, options);

// Export hook for loading requirement by id
export const useRequirement = (id, options) => useInstance(`/api/requirements/${id}/`, options);

// Export hook for loading service by id
export const useService = (id, options) => useInstance(`/api/services/${id}/`, options);

// Export a function for joining a project with an invite code
export const joinProject = async code => apiFetch('/api/join/', { method: 'POST', data: { code } });

// Export hook for loading category by id
export const useCategory = (id, options) => useInstance(`/api/categories/${id}/`, options);

// Export hook for loading resource by id
export const useResourceType = (id, options) => useInstance(`/api/resources/${id}/`, options);


// Export a hook for using a project event stream
export const useProjectEvents = (project, options) => {
    const { fetchPoint = true, autoFetch = true } = options || {};

    const [initialised, setInitialised] = useState(false);
    const [dirty, setDirty] = useState(autoFetch);
    const [fetching, setFetching] = useState(false);
    const [fetchError, setFetchError] = useState(null);
    const [data, setData] = useState([]);

    // Get the events URL from the project links
    const eventsUrl = project.data._links.events;

    // Build the stream object with fetchable methods
    const stream = {
        initialised,
        dirty,
        fetching,
        fetchError,
        data,
        // This function loads more events
        fetch: () => {
            // Build the URL to use
            const url = new URL(eventsUrl);
            // If there is a most recent event, use its time for the since parameter
            const event = data.find(_ => true);
            if( event ) url.searchParams.append("since", event.created_at);
            // Mark the fetch as in progress
            setFetching(true);
            // In order to prevent state updates after a component has been unmounted,
            // we need to be able to cancel fetches
            // To do this, we can use an abort controller
            const controller = new AbortController();
            // Start the fetch using the signal from the controller
            apiFetch(url, { signal: controller.signal })
                // If the promise resolves, update the state with the new data
                .then(newData => {
                    // Prepend the data to the existing data
                    setData([...newData, ...data]);
                    setInitialised(true);
                    setFetching(false);
                    setDirty(false);
                    setFetchError(null);
                })
                // If the promise rejects, update the state with the error
                .catch(error => {
                    // If the error is an abort error, it is because the component that
                    // owns the data has been unmounted so don't update the state
                    if( error.name === "AbortError" ) return;
                    setFetching(false);
                    setDirty(false);
                    setFetchError(error);
                });
            // Return the abort function to allow the fetch to be cancelled from outside
            return () => controller.abort();
        },
        // This function marks the stream as dirty, causing more data to be fetched
        markDirty: () => setDirty(true),
        // This function resets the stream, clearing all the data and triggering a re-fetch
        reset: () => {
            setInitialised(false);
            setDirty(true);
            setFetching(false);
            setFetchError(null);
            setData([]);
        }
    };

    // Hooks cannot be conditional, so we always have to call the fetch point hook
    // However we can effectively disable it by making sure it never sees a dirty stream
    useFetchPoint({ ...stream, dirty: fetchPoint && stream.dirty });

    // We need to be a fetch point for the event stream
    return stream;
};
