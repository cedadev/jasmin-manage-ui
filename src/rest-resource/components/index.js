// Export the delete button and status components directly
export * from './InstanceDeleteButton';
export * from './Status';

// Export the form pieces in their own namespace
import * as Form from './ResourceForm';
export { Form };
