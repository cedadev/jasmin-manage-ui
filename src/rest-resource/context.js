import React, { createContext, useContext, useEffect } from 'react';

import { useEndpoint, useResource, useInstance } from './hooks';


// This higher order function turns a hook into a context, allowing the hook
// state to be shared between components
const createContextForHook = hookFunction => (url, options = {}) => {
    // Make a new context
    const Context = createContext();
    // Return a provider and hook that use the context to share the hook state
    return {
        // The provider is an anchor for the data and also a fetch point
        // However, we do not auto-fetch the data as soon as the provider as mounted
        Provider: ({ children }) => {
            const fetchable = hookFunction(url, { ...options, autoFetch: false });
            return <Context.Provider value={fetchable}>{children}</Context.Provider>;
        },
        // The first time a component using the hook is mounted, the data for the fetchable
        // is initialised if it has not already been initialised
        hook: (autoFetch = true) => {
            const fetchable = useContext(Context);
            useEffect(
                // If the fetchable is not initialised, trigger a fetch by marking it as dirty
                // This should not trigger multiple fetches
                () => { if( autoFetch && !fetchable.initialised ) fetchable.markDirty(); },
                []
            );
            return fetchable;
        }
    };
};


export const createContextForEndpoint = createContextForHook(useEndpoint);
export const createContextForResource = createContextForHook(useResource);
export const createContextForInstance = createContextForHook(useInstance);
