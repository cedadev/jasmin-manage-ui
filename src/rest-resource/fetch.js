import Cookies from 'js-cookie';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, {
    method = 'GET',
    headers = {},
    data,
    body,
    ...options
} = {}) => {
    // Populate the required headers for the request
    const defaultHeaders = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        defaultHeaders['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) defaultHeaders['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, {
        method,
        headers: { ...defaultHeaders, ...headers },
        // If data is given, encode it as JSON and use it as the request body
        // If not, use the given body
        body: data ? JSON.stringify(data) : body,
        // Always include credentials
        credentials: 'include',
        // Forward any other options that were supplied
        ...options
    });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    throw new HttpError(response.status, response.statusText, responseText);
};


// Export method-specific versions of apiFetch
export default apiFetch;
