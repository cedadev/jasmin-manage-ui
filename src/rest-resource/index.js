export * from './components';
export * from './context';
export * from './hooks';

import apiFetch from './fetch';
export { apiFetch };
