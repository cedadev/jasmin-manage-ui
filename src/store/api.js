import Cookies from 'js-cookie';

import { withScope } from './utils';


class HttpError extends Error {
    constructor(status, statusText, responseText) {
        super(statusText);
        this.name = this.constructor.name;
        this.status = status;
        this.statusText = statusText;
        this.responseText = responseText;
    }

    json() {
        return JSON.parse(this.responseText)
    }
}


class InvalidOperationError extends Error {
    constructor(operationName) {
        super(`Invalid operation: ${operationName}`);
        this.name = this.constructor.name;
    }
}


// Utility function that asynchronously waits for the specified duration
const sleep = (ms) => (new Promise(resolve => setTimeout(resolve, ms)));


const apiFetch = async (url, options = {}) => {
    // Populate the required headers for the request
    const headers = {};
    // For POST/PUT/DELETE, declare the content type and include the CSRF token if present
    const method = options.method || 'GET';
    if( ['POST', 'PUT', 'DELETE'].includes(method.toUpperCase()) ) {
        headers['Content-Type'] = 'application/json';
        const csrfToken = Cookies.get('csrftoken');
        if( csrfToken ) headers['X-CSRFToken'] = csrfToken;
    }
    // Make the actual request, injecting the cookie credentials and headers
    const response = await fetch(url, { ...options, headers, credentials: 'include' });
    // For a 204, just return
    if( response.status === 204 ) return;
    // Any other successful response should be JSON
    if( response.ok ) return await response.json();
    // An error response may not be JSON, so read the response as text
    const responseText = await response.text();
    // If the response indicates that the request requires authentication, redirect to auth
    if( [401, 403].includes(response.status) ) {
        // These responses will be JSON and should have a code
        const code = JSON.parse(responseText).code;
        // Some auth schemes use 401 to indicate lack of credentials, and some use
        // 403 with a code of not_authenticated - we need to catch both
        if( response.status === 401 || code === "not_authenticated" ) {
            // Get the path to redirect back to from the current location
            const currentPath = new URL(window.location.href).pathname;
            window.location = `/auth/login/?next=${currentPath}`;
            // Wait for the browser to redirect
            await sleep(2000);
        }
    }
    throw new HttpError(response.status, response.statusText, responseText);
};


const apiGet = (url, options) => apiFetch(url, options);
const apiPost = (url, data, options = {}) => apiFetch(url, {
    ...options,
    method: 'POST',
    body: data ? JSON.stringify(data) : undefined
});
const apiPut = (url, data, options = {}) => apiFetch(url, {
    ...options,
    method: 'PUT',
    body: data ? JSON.stringify(data) : undefined
});
const apiDelete = (url, options = {}) => apiFetch(url, {
    ...options,
    method: 'DELETE'
});


export const newEndpoint = path => (mutate, getState) => ({
    // This is true once the endpoint has been fetched once
    initialised: false,
    // Indicates whether a fetch is currently in progress
    fetching: false,
    // The error from the last fetch, if there is one
    fetchError: null,
    initialise: async () => {
        const resource = getState();
        // If the resource is not initialised, trigger a fetch
        if( !resource.initialised ) await resource.fetch();
    },
    fetch: async () => {
        // If there is a fetch in progress, wait for the fetch to finish in a non-blocking way
        if( getState().fetching ) {
            while( getState().fetching ) await sleep(1);
            return;
        }
        // Otherwise, initiate a fetch
        mutate(state => { state.fetching = true; });
        try {
            // Fetch the resources from the API
            const result = await apiGet(path);
            // Update the state
            mutate(state => {
                // We are no longer fetching
                state.fetching = false;
                // We are also initialised
                state.initialised = true;
                // We also did a fetch without errors
                state.fetchError = null;
                // Just store the data as-is
                state.data = result;
            });
        }
        catch(error) {
            // For an error response, mark the fetch as over and store the error
            mutate(state => {
                state.fetching = false;
                state.fetchError = error;
            });
            // Rethrow the error
            throw error;
        }
    }
});


const defaultResource = {
    // This is true once there has been at least one successful fetch
    initialised: false,
    // Indicates whether a fetch is currently in progress
    fetching: false,
    // The error from the last fetch, if there is one
    fetchError: null,
    // Indicates whether a create is currently in progress
    creating: false,
    // The current data as a map of id => instance
    data: {},
    // By default, initialise is a no-op
    initialise: async () => {/* NOOP */},
    // By default, all other operations are not valid
    fetch: async () => { throw new InvalidOperationError('fetch'); },
    create: async () => { throw new InvalidOperationError('create'); },
    aggregate: () => { throw new InvalidOperationError('aggregate'); }
};


export const newResource = resourcePath => (mutate, getState, getParent) => ({
    ...defaultResource,
    // If this is a nested resource, this will return the parent resource
    parent: () => (getParent ? getParent() : undefined),
    // Initialises the resource
    initialise: async () => {
        const resource = getState();
        // If the resource is not initialised, trigger a fetch
        if( !resource.initialised ) await resource.fetch();
    },
    // Fetches all the instances for the resource
    fetch: async () => {
        // If there is a fetch in progress, wait for the fetch to finish in a non-blocking way
        if( getState().fetching ) {
            while( getState().fetching ) await sleep(1);
            return;
        }
        // Otherwise, initiate a fetch
        mutate(state => { state.fetching = true; });
        try {
            // Fetch the resources from the API
            const resourceList = await apiGet(resourcePath);
            // Update the state
            mutate(state => {
                // We are no longer fetching
                state.fetching = false;
                // We are also initialised
                state.initialised = true;
                // We also did a fetch without errors
                state.fetchError = null;
                // The data is stored as a dictionary of instances
                // Create or update the instance for each item in the returned list
                resourceList.forEach(resource => {
                    state.data[resource.id] = state.data.hasOwnProperty(resource.id) ?
                        // If there is an existing instance, update it with the new data
                        Object.assign(state.data[resource.id].data, resource) :
                        // If there is no existing instance, make a new one
                        newInstance(resource)(...withScope(mutate, getState, ["data", resource.id]));
                });
                // Remove items that are no longer in the list
                //   (remember that Javascript objects implicitly convert keys to strings)
                const nextIds = resourceList.map(resource => resource.id.toString());
                Object.keys(state.data).filter(id => !nextIds.includes(id)).forEach(id => { delete state.data[id]; });
            });
        }
        catch(error) {
            // For an error response, mark the fetch as over and store the error
            mutate(state => {
                state.fetching = false;
                state.fetchError = error;
            });
            // Rethrow the error
            throw error;
        }
    },
    // Creates a new instance of the resource
    create: async data => {
        // Mark the create as in progress
        mutate(state => { state.creating = true; });
        try {
            // Attempt to create the resource
            const resource = await apiPost(resourcePath, data);
            // Update the state if the create was successful
            mutate(state => {
                // We are no longer creating
                state.creating = false;
                // Add the new resource to the data
                state.data[resource.id] = newInstance(resource)(...withScope(mutate, getState, ["data", resource.id]));
            });
        }
        catch(error) {
            // Mark the create as over before re-throwing the error
            mutate(state => { state.creating = false; });
            throw error;
        }
    },
    // Initialise nested resources to an arbitrary depth using the URLs from links
    initNested: async (...keys) => {
        // Before the nested resources can be initialised, we need to be initialised
        await getState().initialise();
        // If no keys are given, there is nothing more to do
        if( keys.length === 0 ) return;
        // Otherwise, initialise the nested resources for each instance that was loaded
        const [firstKey, ...otherKeys] = keys;
        await Promise.all(
            Object.values(getState().data).map(instance =>
                instance.nested(firstKey).initNested(...otherKeys)
            )
        );
    },
    // Returns an aggregated resource across several levels of nesting
    aggregate: (...keys) => {
        const resource = getState();
        // If there are no keys, the aggregate resource is this resource
        if( keys.length === 0 ) return resource;
        // If this resource is not initialised, then there is nothing to aggregate
        // But the aggregate resource should also be considered not initialised
        if( !resource.initialised ) {
            return {
                ...defaultResource,
                // Initialising the aggregate means initialising the nested resources
                initialise: () => getState().initNested(...keys)
            };
        }
        // For each item in our data, use the first key to get a nested resource then
        // aggregate the remaining keys over the nested resource
        // Then collapse those resources to get the aggregate state
        const [firstKey, ...otherKeys] = keys;
        const aggregateState = Object.values(resource.data)
            .map(instance => instance.nested(firstKey).aggregate(...otherKeys))
            .reduce(
                (state1, state2) => ({
                    ...state1,
                    initialised: state1.initialised && state2.initialised,
                    fetching: state1.fetching || state2.fetching,
                    fetchError: state1.fetchError || state2.fetchError,
                    data: { ...state1.data, ...state2.data }
                }),
                // If there are no aggregates, we are initialised
                { ...defaultResource, initialised: true }
            );
        // Return a resource - aggregate resources only support initialise
        return {
            ...aggregateState,
            // Initialising the aggregate means initialising the nested resources
            initialise: () => getState().initNested(...keys)
        };
    }
});


export const newInstance = (initialData) => (mutate, getState, getParent) => ({
    // Indicates if the instance is currently being fetched
    fetching: false,
    // Indicates if the instance is currently being updated
    updating: false,
    // Indicates if the instance is currently being deleted
    deleting: false,
    // The current data for the instance
    data: initialData,
    // Initialise a nested resource for each key in the _links (that isn't self)
    // even if they aren't actually nested resources
    // Lazily loading these on demand is a major pain
    nestedResources: Object.assign(
        {},
        ...Object.entries(initialData._links)
            .filter(([name, _]) => !["self", "_links"].includes(name))
            .map(([name, url]) => ({
                [name]: newResource(url)(
                    // The resource requires scoped mutate and getState
                    ...withScope(mutate, getState, ["nestedResources", name])
                )
            }))
    ),
    // Returns the resource that the instance belongs to
    resource: getParent,
    // Returns a named nested resource
    nested: name => getState().nestedResources[name],
    // Update the instance with new data
    update: async data => {
        // Mark the update as in progress
        mutate(state => { state.updating = true; });
        try {
            const selfUrl = getState().data._links.self;
            // Attempt to update the resource
            const resource = await apiPut(selfUrl, data);
            // Update the state if the update was successful
            mutate(state => {
                state.updating = false;
                Object.assign(state.data, resource);
            });
        }
        catch(error) {
            // Mark the update as over before re-throwing the error
            mutate(state => { state.updating = false; });
            throw error;
        }
    },
    // Delete the instance
    delete: async () => {
        // Mark the delete as in progress
        mutate(state => { state.deleting = true; });
        try {
            const { id, _links: { self: selfUrl } } = getState().data;
            // Attempt to delete the resource
            await apiDelete(selfUrl);
            // Update the state if the delete was successful
            // We have to remove the instance from the parent resource
            mutate((state, parent) => {
                state.deleting = false;
                delete parent.data[id];
            });
        }
        catch(error) {
            // Mark the delete as over before re-throwing the error
            mutate(state => { state.deleting = false; });
            throw error;
        }
    },
    // Executes a named action on the resource
    executeAction: async (name, data) => {
        // Mark the resource as updating
        mutate(state => { state.updating = true; });
        try {
            // Get the URL for the action
            const actionUrl = getState().data._links[name];
            // Attempt to create the resource
            const resource = await apiPost(actionUrl, data);
            // Update the state if the action was successful
            mutate(state => {
                // We are no longer updating
                state.updating = false;
                // Update the resource data
                Object.assign(state.data, resource);
            });
        }
        catch(error) {
            // Mark the update as over before re-throwing the error
            mutate(state => { state.updating = false; });
            throw error;
        }
    }
});
