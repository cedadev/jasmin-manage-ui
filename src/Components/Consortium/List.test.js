// src/Components/Consortium/List.test.js

import React from "react";
import { render, screen, waitFor, within } from "@testing-library/react";
import renderer from "react-test-renderer";

import ConsortiumList from "./List";
import { useCurrentUser, useConsortia } from "../../api";
import { BrowserRouter as Router } from "react-router-dom";
import { useEnsureRefreshed } from "../../rest-resource";

// Mock the API hooks
jest.mock("../../api");

// Mock the useEnsureRefreshed hook
jest.mock("../../rest-resource", () => ({
  ...jest.requireActual("../../rest-resource"),
  useEnsureRefreshed: jest.fn(),
}));

const renderPage = () => {
  return render(
    <Router>
      <ConsortiumList />
    </Router>
  );
};

describe("ConsortiumList Component", () => {
  const mockConsortia = [
    {
      data: {
        id: 1,
        name: "Consortium A",
        description: "Description A",
        is_public: true,
        num_projects: 5,
        num_projects_current_user: 2,
        manager: {
          id: 1,
          first_name: "Alice",
          last_name: "Smith",
          username: "alice",
        },
      },
    },
    {
      data: {
        id: 2,
        name: "Consortium B",
        description: "Description B",
        is_public: false,
        num_projects: 1,
        num_projects_current_user: 0,
        manager: {
          id: 2,
          first_name: "Bob",
          last_name: "Jones",
          username: "bob",
        },
      },
    },
  ];

  const mockCurrentUser = {
    data: {
      id: 1,
      first_name: "Alice",
      last_name: "Smith",
      username: "alice",
    },
  };

  beforeEach(() => {
    useCurrentUser.mockReturnValue(mockCurrentUser);
    useConsortia.mockReturnValue({
      data: {
        1: mockConsortia[0],
        2: mockConsortia[1],
      },
      initialised: true,
      fetching: false,
      fetchError: null,
      markDirty: jest.fn(),
    });

    useEnsureRefreshed.mockImplementation((consortia) => consortia);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("matches the snapshot", () => {
    const component = renderer.create(
      <Router>
        <ConsortiumList />
      </Router>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders loading state", () => {
    useConsortia.mockReturnValue({
      initialised: false,
      fetching: true,
      markDirty: jest.fn(),
    });
    renderPage();

    expect(screen.getByText("Loading consortia...")).toBeInTheDocument();
  });

  it("renders error state", () => {
    useConsortia.mockReturnValue({
      data: null,
      initialised: false,
      fetching: false,
      fetchError: new Error("Failed to load consortia"),
    });

    renderPage();

    expect(screen.getByText("Unable to load consortia.")).toBeInTheDocument();
    expect(screen.getByRole("alert")).toHaveClass("alert-danger");
  });

  it("renders consortia list", async () => {
    renderPage();

    await waitFor(() => {
      expect(screen.getByText("Consortia")).toBeInTheDocument();
    });

    const consortiumAHeader = screen.getByText("Consortium A");
    const consortiumACard = consortiumAHeader.closest(".card");
    const consortiumBHeader = screen.getByText("Consortium B");
    const consortiumBCard = consortiumBHeader.closest(".card");

    const withinConsortiumA = within(consortiumACard);
    expect(withinConsortiumA.getByText("Description A")).toBeInTheDocument();
    expect(withinConsortiumA.getByText(/Consortium has/)).toBeInTheDocument();
    expect(withinConsortiumA.getByText(/5 projects/)).toBeInTheDocument();

    expect(withinConsortiumA.getByText(/You have/)).toBeInTheDocument();
    expect(withinConsortiumA.getByText(/2 projects/)).toBeInTheDocument();
    expect(
      withinConsortiumA.getByText(/in this consortium./)
    ).toBeInTheDocument();
    expect(
      withinConsortiumA.getByText("You are the consortium manager.")
    ).toBeInTheDocument();
    expect(withinConsortiumA.getByText("Go to consortium")).toBeInTheDocument();

    const withinConsortiumB = within(consortiumBCard);
    expect(withinConsortiumB.getByText("Description B")).toBeInTheDocument();
    expect(
      withinConsortiumB.getByText(/This consortium is/)
    ).toBeInTheDocument();
    expect(withinConsortiumB.getByText(/not public/)).toBeInTheDocument();
    expect(withinConsortiumB.getByText(/Consortium has/)).toBeInTheDocument();
    expect(withinConsortiumB.getByText(/1 project/)).toBeInTheDocument();
    expect(withinConsortiumB.getByText(/Manager is/)).toBeInTheDocument();
    expect(withinConsortiumB.getByText(/Bob Jones/)).toBeInTheDocument();
  });

  it("renders manager's username when last name is not available", () => {
    useCurrentUser.mockReturnValue({
      data: {
        id: 2,
        first_name: "Test",
        last_name: "User",
        username: "testuser",
      },
      fetching: false,
      fetchError: null,
      initialised: true,
    });

    useConsortia.mockReturnValue({
      data: {
        1: {
          data: {
            id: 1,
            name: "Consortium 2",
            description: "Description 2",
            num_projects: 2,
            num_projects_current_user: 1,
            manager: {
              id: 1,
              username: "manageruser",
              first_name: "",
              last_name: null,
            },
            is_public: true,
          },
        },
      },
      initialised: true,
      fetching: false,
      fetchError: null,
    });

    renderPage();

    expect(screen.getByText(/Manager is/)).toBeInTheDocument();
    expect(screen.getByText(/manageruser/)).toBeInTheDocument();
  });
});
