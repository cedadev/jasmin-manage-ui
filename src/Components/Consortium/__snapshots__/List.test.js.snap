// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConsortiumList Component matches the snapshot 1`] = `
[
  <div
    className="row"
  >
    <div
      className="col"
    >
      <h1
        className="border-bottom mt-4 mb-3"
      >
        Consortia
      </h1>
    </div>
  </div>,
  <div
    className="row row-cols-lg-3 row-cols-sm-2 row-cols-1"
  >
    <div
      className="col"
    >
      <div
        className="mb-3 card"
        style={
          {
            "borderWidth": "3px",
          }
        }
      >
        <div
          className="card-header"
        >
          <h5
            className="mb-0"
          >
            Consortium A
          </h5>
        </div>
        <div
          className="border-0 list-group list-group-flush"
          onKeyDown={[Function]}
        >
          <div
            className="list-group-item"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            Description A
          </div>
          <div
            className="list-group-item"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            Consortium has
             
            <strong>
              5
               project
              s
            </strong>
            .
          </div>
          <div
            className="list-group-item list-group-item-warning"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            You have
             
            <strong>
              2
               project
              s
            </strong>
             
            in this consortium.
          </div>
          <div
            className="list-group-item list-group-item-success"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            <strong>
              You are the consortium manager.
            </strong>
          </div>
        </div>
        <div
          className="text-right card-footer"
        >
          <a
            href="/consortia/1"
            onClick={[Function]}
          >
            <button
              className="btn btn-outline-primary"
              disabled={false}
              type="button"
            >
              Go to consortium
            </button>
          </a>
        </div>
      </div>
    </div>
    <div
      className="col"
    >
      <div
        className="mb-3 card"
        style={
          {
            "borderWidth": "3px",
          }
        }
      >
        <div
          className="card-header"
        >
          <h5
            className="mb-0"
          >
            Consortium B
          </h5>
        </div>
        <div
          className="border-0 list-group list-group-flush"
          onKeyDown={[Function]}
        >
          <div
            className="list-group-item"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            Description B
          </div>
          <div
            className="list-group-item list-group-item-danger"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            This consortium is 
            <strong>
              not public
            </strong>
            .
          </div>
          <div
            className="list-group-item"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            Consortium has
             
            <strong>
              1
               project
              
            </strong>
            .
          </div>
          <div
            className="list-group-item"
            data-rb-event-key={null}
            disabled={false}
            onClick={[Function]}
          >
            Manager is 
            <strong>
              Bob Jones
            </strong>
            .
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
